model/device_deployment.go

```diff
@@ -17,119 +17,143 @@ package model
 )
 
+var allStatuses = []interface{}{ // NOTE: []DeviceDeploymentStatus
+	DeviceDeploymentStatusNoArtifact,
+	DeviceDeploymentStatusFailure,
+	DeviceDeploymentStatusSuccess,
+	DeviceDeploymentStatusPending,
+	DeviceDeploymentStatusRebooting,
+	DeviceDeploymentStatusInstalling,
+	DeviceDeploymentStatusDownloading,
+	DeviceDeploymentStatusAlreadyInst,
+	DeviceDeploymentStatusAborted,
+	DeviceDeploymentStatusDecommissioned,
+}
+
+func (stat DeviceDeploymentStatus) Validate() error {
+	return validation.In(allStatuses...).
+		Validate(stat)
+}
+
 // DeviceDeploymentStatus is a helper type for reporting status changes through
 // the layers
-type DeviceDeploymentStatus struct {
+type DeviceDeploymentState struct {
 	// status reported by device
-	Status string `valid:"required"`
+	Status DeviceDeploymentStatus
 	// substate reported by device
-	SubState *string
+	SubState string `json:",omitempty" bson:",omitempty"`
 	// finish time
-	FinishTime *time.Time
+	FinishTime *time.Time `json:",omitempty" bson:",omitempty"`
+}
+
+func (state DeviceDeploymentState) Validate() error {
+	return validation.ValidateStruct(&state,
+		validation.Field(&state.Status, validation.Required),
+	)
 }
 
 type DeviceDeployment struct {
 	// Internal field of initial creation of deployment
-	Created *time.Time `json:"created" valid:"required"`
+	Created *time.Time `json:"created" bson:"created"`
 
 	// Update finish time
-	Finished *time.Time `json:"finished,omitempty" valid:"-"`
+	Finished *time.Time `json:"finished,omitempty" bson:"finished,omitempty"`
 
 	// Status
-	Status *string `json:"status" valid:"required"`
+	Status DeviceDeploymentStatus `json:"status" bson:"status"`
 
 	// Device id
-	DeviceId *string `json:"id" valid:"required"`
+	DeviceId string `json:"id" bson:"deviceid"`
 
 	// Deployment id
-	DeploymentId *string `json:"-" valid:"uuidv4,required"`
+	DeploymentId string `json:"-" bson:"deploymentid"`
 
 	// ID
-	Id *string `json:"-" bson:"_id" valid:"uuidv4,required"`
+	Id string `json:"-" bson:"_id"`
 
 	// Assigned software image
-	Image *Image `json:"-" valid:"-"`
+	Image *Image `json:"-"`
 
 	// Target device type
-	DeviceType *string `json:"device_type,omitempty" valid:"-"`
+	DeviceType string `json:"device_type,omitempty" bson:"devicetype"`
 
 	// Presence of deployment log
-	IsLogAvailable bool `json:"log" valid:"-" bson:"log"`
+	IsLogAvailable bool `json:"log" bson:"log"`
 
 	// Device reported substate
-	SubState *string `json:"substate,omitempty" valid:"-" bson:"substate"`
+	SubState string `json:"substate,omitempty" bson:"substate,omitempty"`
 }
 
-func NewDeviceDeployment(deviceId, deploymentId string) (*DeviceDeployment, error) {
+func NewDeviceDeployment(deviceId, deploymentId string) *DeviceDeployment {
 
 	now := time.Now()
-	initStatus := DeviceDeploymentStatusPending
 
-	uid := uuid.NewV4()
+	uid, err := uuid.NewRandom()
+	if err != nil {
+		panic(errors.Wrap(err, "failed to generate random uuid (v4)"))
```

Comment:
I'm not sure how it will be handled by the go-json-rest library
I think 500 will be sent, which is good, but without the body, which may not be good, because in all other cases we are sending an error message with http 500 responses

Reply:
The only way this can happen is if the uuid library fails to read from the RNG (which I only think can happen if the machine runs out of entropy). Either way I don't think we should expose this information to the API and let the recovery middleware push a the stacktrace to the logs.